### ISSUE TITLE
Use L-SVRGDA to minimize a function

### ISSUE THREAD

**Issue by MidouDali** [hugobb/sgda#1]
Hi,
I want to use the L-SVRGDA method to minimize a function which depends on n variables (tensors). Can you help me how to use it?
thank you

**Comment by MidouDali** [comment]
assuming that I will calculate the values of x and y in each iteration to minimize the function f (figure). What's the update rule to use? Can you please simplify the method L-SVRGDA with this example.
thank you
<img width="155" alt="1" src="https://user-images.githubusercontent.com/21183538/213319690-76a95c11-1556-4b78-ac53-701ec23838dc.png">

**Comment by hugobb** [comment]
L-SVRGDA is an algorithm to solve min-max problems of the form: $$\min_x \max_y f(x, y)$$
If you just want to minimize a function I recommend you to use L-SVRG ([see Algorithm 1](https://proceedings.mlr.press/v117/kovalev20a/kovalev20a.pdf#page=5)).
Note that in order to you use those algorithms you need to be able to compute the gradient of $f$.

**Comment by MidouDali** [comment]
Thank you,
I want to minimize a function f 
<img width="197" alt="2" src="https://user-images.githubusercontent.com/21183538/213504145-5d177f92-7aeb-4fad-9a34-8ad3daba900d.png">

Yes, I am able to compute the gradient. Can you simplify the L-SVRG algorithm with ana example and help me to apply it.
thank you

**Comment by hugobb** [comment]
Are you familiar with SGD (Stochastic Gradient Descent) or Gradient Descent ? if not I recommend you to read about it first before getting familiar with L-SVRG. L-SVRG is an extension of SGD that use a variance-reduction technique to reduce noise in the gradient and improve convergence.
Here is a list of references to help you:
- https://optimization.cbe.cornell.edu/index.php?title=Stochastic_gradient_descent
- https://harkiratbehl.github.io/projects/svrg/svrg.pdf
- https://www.cs.cmu.edu/~pradeepr/convexopt/Lecture_Slides/stochastic_optimization_methods.pdf

If you already familiar with SGD, can you clarify exactly what you don't understand about the description of L-SVRG ?

**Comment by MidouDali** [comment]
Thank you,
I don't understand what's the difference between 
<img width="43" alt="gr1" src="https://user-images.githubusercontent.com/21183538/213524055-54cc2786-73e0-4c4f-a87f-13053f9dde4c.png">
and 
<img width="42" alt="gr2" src="https://user-images.githubusercontent.com/21183538/213524074-36ee3e76-2ebf-4ca3-a31d-b0da468ab262.png">
What I understand that the seond is a full gradient. Can I compute it from other gradients?
I don't understand also how to update the reference point wk, what's the meaning of this?: 
<img width="233" alt="probab2" src="https://user-images.githubusercontent.com/21183538/213524863-03861619-5b17-4449-9522-ac30d85adba7.png">

**Comment by hugobb** [comment]
Let me try to clarify this: the idea behind L-SVRG is to reduce the variance of the stochastic gradient by keeping a "memory" of a previous iteration. 

Let's first take a look at:
![image](https://user-images.githubusercontent.com/22268620/213529192-e823720a-d474-4f97-bc39-237a89b58389.png)
Here we simply update the "memory" $w$ with probability $p$. Thus with probability $p$ we update the "memory" with the current iteration, otherwise we don't do update the memory.

Everytime we update the "memory", we also compute the full batch gradient associated with it: $$\nabla f(w^k) = \sum_i \nabla f_i(w^k)$$
This is expensive but we usually choose $p$ to be quite small so we don't have to do this often.

To clarify a bit with some code:
```python
import random

def f(x, i):
    """The function we want to minimize"""


def grad_f(x, i):
    """The stochastic gradient of the function we want to minimize"""


def full_grad_f(x):
    """The full gradient of the function we want to minimize"""
    return sum(grad_f(w, i) for i in range(n)) / n

lr = 0.1  # learning rate
p = 0.1  # probability of updating the memory
K = 100  # number of iterations
n = 10  # number of samples
x = ...  # initial value for x
w = x
full_grad_f_w = full_grad_f(w)
for k in range(K):
    i = random.randrange(n)
    g_k = grad_f(x, i) - grad_f(w, i) + full_grad_f_w
    x = x - lr * g_k
    if random.random() < p:
        w = x
        full_grad_f_w = full_grad_f(w)

```

**Comment by MidouDali** [comment]
thank you very much, this is very helpful
The function f has 4 variables: x, y, z and q
Do you mean by samples, the elements in a batch of data while training or the variables x, y, z and q?
this algorithm can be applied for every variable x, y, z and q?
thank you

**Comment by hugobb** [comment]
By samples I mean the elements in a batch of data. Yes in my example x can be a vector that contains several variables.

**Comment by MidouDali** [comment]
Hi,
I am implementing a code using L-SVRG. I have 2 questions please :
- to update the reference w, we use x(k) or x(k+1)? In your code, the new value of w must be the new value of x (x(k+1)), but in the paper w= x(k)
<img width="490" alt="1" src="https://user-images.githubusercontent.com/21183538/215772228-d0e2410b-40c9-4f37-8b55-3604db1cfc4b.png">
- suppose we updated w and computed the full_grad_f_w associated  with it during iteration i. At an iteration (i+t), do we recalculate full_grad_f_w according to the data of this iteration or use its old stored value (computed according to the data of iteration i)?
thank you

**Comment by hugobb** [comment]
- From my understanding there is no difference between using $x^k$ or $x^{k+1}$ when updating $w$
- When we calculate full_grad_f_w we can store it in memory and re-use it at every iteration. We only recompute it when $w$ gets updated.
